src/kernel/bitcoinkernel.h

```diff
@@ -0,0 +1,1228 @@
+ *
+ * These are eventually placed into a kernel context through the kernel context
+ * options. The parameters describe the properties of a chain, and may be
+ * instantiated for either mainnet, testnet, signet, or regtest.
+ */
+typedef struct kernel_ChainParameters kernel_ChainParameters;
+
+/**
+ * Opaque data structure for holding options for creating a new kernel context.
+ *
+ * Once a kernel context has been created from these options, they may be
+ * destroyed. The options hold the notification callbacks as well as the
+ * selected chain type until they are passed to the context. If no options are
+ * configured, the context will be instantiated with no callbacks and for
+ * mainnet. Their content and scope can be expanded over time.
+ */
+typedef struct kernel_ContextOptions kernel_ContextOptions;
+
+/**
+ * Opaque data structure for holding a kernel context.
+ *
+ * The kernel context is used to initialize internal state and hold the chain
+ * parameters and callbacks for handling error and validation events. Once other
+ * validation objects are instantiated from it, the context needs to be kept in
+ * memory for the duration of their lifetimes.
+ *
+ * The processing of validation events is done through an internal task runner
+ * owned by the context. It passes events through the registered validation
+ * interface callbacks.
+ *
+ * A constructed context can be safely used from multiple threads.
+ */
+typedef struct kernel_Context kernel_Context;
+
+/**
+ * Opaque data structure for holding a block index pointer.
+ *
+ * This is a pointer to an element in the block index currently in memory of the
+ * chainstate manager. It is valid for the lifetime of the chainstate manager it
+ * was retrieved from.
+ */
+typedef struct kernel_BlockIndex kernel_BlockIndex;
+
+/**
+ * Opaque data structure for holding options for creating a new chainstate
+ * manager.
+ *
+ * The chainstate manager options are used to set some parameters for the
+ * chainstate manager. For now it just holds default options.
+ */
+typedef struct kernel_ChainstateManagerOptions kernel_ChainstateManagerOptions;
+
+/**
+ * Opaque data structure for holding a chainstate manager.
+ *
+ * The chainstate manager is the central object for doing validation tasks as
+ * well as retrieving data from the chain. Internally it is a complex data
+ * structure with diverse functionality.
+ *
+ * The chainstate manager is only valid for as long as the context with which it
+ * was created remains in memory.
+ *
+ * Its functionality will be more and more exposed in the future.
+ */
+typedef struct kernel_ChainstateManager kernel_ChainstateManager;
+
+/**
+ * Opaque data structure for holding a block.
+ */
+typedef struct kernel_Block kernel_Block;
+
+/**
+ * Opaque data structure for holding a non-owned block. This is typically a
+ * block available to the user through one of the validation callbacks.
+ */
+typedef struct kernel_BlockPointer kernel_BlockPointer;
```

Comment:
We need `kernel_BlockPointer` because the validation interface gives us a non-owning reference. It would imo be a lot nicer if we could generalize this into `kernel_Block`, so I have opened #33078 to improve ownership semantics in the validation interface, after which `kernel_BlockPointer` can then be removed, e.g. as in https://github.com/stickies-v/bitcoin/commits/kernel/remove-blockpointer/